### PR TITLE
Reservoir RFC Water Elevation Tracking

### PR DESCRIPTION
**Pull request by jdmattern**
TYPE: enhancement

KEYWORDS: reservoirs, water management, water elevation

SOURCE: David Mattern, NOAA-OWP, @jdmattern-noaa 

DESCRIPTION OF CHANGES: Updates isolated to reservoir module. Updates to how RFC type reservoirs balance and ensure boundary conditions of water elevation from the assimilated RFC forecasted outflows.

ISSUE: 
```
None
```

TESTS CONDUCTED: Reservoir unit tests updated and passed. 

NOTES: Targeted for NWM V3.0

### Checklist
Merging the PR depends on following checklist being completed. Add `X` between each of the square 
brackets if they are completed in the PR itself. If a bullet is not relevant to you, please comment 
on why below the bullet.

 - [ ] Closes issue #xxxx (An issue must exist or be created to be closed. The issue describes and documents the problem and general solution, the PR describes the technical details of the solution.) 
 - [x] Tests added (unit tests and/or regression/integration tests)
 - [x] Backwards compatible
 - [ ] Requires new files? If so, how to generate them.
 - [x] Fully documented
 - [ ] Short description in the Development section of `NEWS.md`

### CODE REVIEW

File: trunk/NDHMS/Routing/Reservoirs/RFC_Forecasts/module_rfc_forecasts.F [hunk 1]
```diff
@@ -310,6 +311,19 @@ subroutine run_rfc_forecasts_reservoir(this, previous_timestep_inflow, inflow, &
             ! Set outflow to corresponding discharge from array
             this%output%outflow = this%state%discharges(this%state%time_series_index)
 
+            ! Update water elevation
+            this%state%water_elevation = this%state%water_elevation + &
+            ((this%input%inflow - this%output%outflow) / this%properties%lake_area) * routing_period
+
+            ! Ensure that the water elevation is within the minimum and maximum elevation
+            if (this%state%water_elevation < 0.0) then
+                this%state%water_elevation = 0.0
```
Comment (donaldwj) [inline trunk/NDHMS/Routing/Reservoirs/RFC_Forecasts/module_rfc_forecasts.F:320]: Is water elevation height above lake bed or height above sea level? in the second case this may be invalid in a couple specific areas (Death Valley, New Orleans, Denmark)

**Review by donaldwj — DISMISSED** [review]
The only issue I see is the change to force level pool elevation to max elevation will cause mass balance errors. This is probably fine as if a lake or levee goes over bank mass will be lost (and maybe the dam also). We may want to account for the excess mass to eventually do something with it (either force into touching terrain grids or backwater forcing on the upstream)

**Comment by scrasmussen** [comment]
When starting the merge process there are conflicts with the precision of the integer `lake_number` that has cascading effects in the code that needs to be changed. The variable `integer :: lake_number` will need to be updated to `integer(kind=int64)` in a number of places. I'm working on that right.

**Review by rcabell — APPROVED** [review]
CI passes on Croton and CONUS, and merged reservoir Unit Tests pass on Cheyenne. Recommend merge.

**Review by scrasmussen — APPROVED** [review]
LGTM, nice job with the rebase @rcabell!
